### ISSUE TITLE
Import duplicate transaction failed to be detected

### ISSUE THREAD

**Issue by MeRoBo** [firefly-iii/firefly-iii#2358]
**Bug description**
I am running Firefly III version 4.7.17, and my problem is:

When importing transactions from the bank with overlap of the previous export, the duplicate transaction detection only detects some transactions. On redacting the import errors I have notice only credits duplicates are being detected.

**Steps to reproduce**

Imported 3 weeks of transactions from the bank. A week later I imported the last 30 days (overlaping 3 weeks) and only some (non sequential rows) reported as duplicate causing major disruptions to the balances. 

**Expected behavior**
Detect which transactions have been imported.
- create a hash token of csv row that is being imported) 
- check the hash token doesn't already exist in the DB transaction table
- if not found record transaction with hash for future lookup
- if found record as duplicate and dismiss.

**Extra info**
Debug information generated at 2019-07-31 08:59:25 UTC for Firefly III version **4.7.17**.

| Variable | Content |
| --- | --- |
| FF version | 4.7.17 |
| FF API version | 0.9.2 |
| App environment | local |
| App debug mode | '' |
| App cache driver | file |
| App logging | , stdout |
| PHP version | 7.2.16 |
| Display errors | Off |
| Session start | 2019-07-01 00:00:00 |
| Session end | 2019-07-31 23:59:59 |
| Session first | 2018-09-11 00:00:00 |
| Error reporting | ALL errors |
| Host | Linux |
| Interface | apache2handler |
| UserID | 1 |
| Attempt at "en" | false |
| Attempt at "English" | false |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Login provider |   |
| Storage disks | local-upload  |
| Using Sandstorm? | no |
| Is Sandstorm (.env) | false |
| Is Docker (.env) | true |
| bunq uses sandbox | false |
| Trusted proxies (.env) |  |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36 |
| Loaded extensions | Core, date, libxml, openssl, pcre, sqlite3, zlib, ctype, curl, dom, fileinfo, filter, ftp, hash, iconv, json, mbstring, SPL, PDO, session, posix, Reflection, standard, SimpleXML, pdo_sqlite, Phar, tokenizer, xml, xmlreader, xmlwriter, mysqlnd, apache2handler, bcmath, gd, intl, ldap, memcached, pdo_mysql, pdo_pgsql, sodium, zip, Zend OPcache |

**Bonus points**
Import Errors recorded in import_jobs table.

- "Row #25 (\"WOOLWORTHS ONLINE 8843 - Visa Refund REDACTED\") could not be imported. It already exists.",
- "Row #30 (\"WOOLWORTHS ONLINE 8843 - Visa Refund REDACTED\") could not be imported. It already exists.",
- "Row #38 (\"Salary - Salary Deposit - REDACTED\") could not be imported. It already exists.",
- "Row #40 (\"Internal Transfer - Internal Transfer - REDACTED\") could not be imported. It already exists.",
- "Row #41 (\"Internal Transfer - Internal Transfer - REDACTED\") could not be imported. It already exists.",
- "Row #51 (\"WOOLWORTHS ONLINE 8843 - Visa Refund - REDACTED\") could not be imported. It already exists.",
- "Row #60 (\"Salary - Receipt REDACTED\") could not be imported. It already exists.",
- "Row #62 (\"WOOLWORTHS ONLINE 8843 - REDACTED\") could not be imported. It already exists.",
- "Row #64 (\"WOOLWORTHS ONLINE 8843 - REDACTED\") could not be imported. It already exists.",
- "Row #81 (\"Salary - Salary Deposit - REDACTED\") could not be imported. It already exists.",
- "Row #87 (\"WOOLWORTHS ONLINE 8843 - Visa Refund REDACTED\") could not be imported. It already exists.",
- "Row #94 (\"Salary - Receipt REDACTED\") could not be imported. It already exists."
-

**Comment by JC5** [comment]
In order to reproduce this I'll need some real life reproduceable examples, or the log files for the same.

**Comment by JC5** [comment]
Hi! This is a canned reply that tells you how to get more debug details.

1. Please collect some log files. Go to `storage/logs` and open the latest log file. It should be called something like "ff3-fpm-fcgi-*.log". In it, you should be able to find the error along with its "stacktrace": this will tell me where Firefly III broke down. If you run Docker, use `docker logs <container>`.
2. You may also [enable debug mode](https://github.com/firefly-iii/help/wiki/Enable-debug-logging-and-debug-mode) to make sure the log files catch everything and try again.
3. Make sure you clear your browser cookies and clear Firefly III's cache files by running `php artisan cache:clear` or by visiting `/flush` on your installation.

If you do not want to share log files or other details on Github, please [send them to me over (encrypted) email](https://firefly-iii.org/contact-information.html).

**Comment by MeRoBo** [comment]
Hi James,

I can replicate this by restoring my DB, flushing the cache, then importing
the overlap csv file.

I found that importing without applying the rules produced an unknown
result, however importing with applying the rules caused duplicate
transactions.

Since you needed logs I tried creating obfuscated versions of my imports
and see if I could reproduce the issue.

I tried all combinations and it seem to work as you claim.

I then tried with my real data on a vanilla install and it worked as you
claimed.

So there must be something wrong with my db state so I will just restart
the imports again from a vanilla state.

In saying that, it would be nice to export/import Bills, rules and budgets.

Thanks for your time.
